### PR TITLE
Refactor Test Users API

### PR DESCRIPTION
**Pull request by uyen3893**
Please review it

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
     console.log(`App running on port ${port}.`)
   })
   //API of users
-  app.get('/users', db.getUsers)
+  app.get('/users', db.getUsersAsync)
+  //app.get('/users', db.getUsers)
   app.get('/users/:id', db.getUserByID)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 2]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
+  // app.get('/users/:id', db.getUserByIDAsync)
   app.post('/users', db.createUser)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 3]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
+  //app.put('/users/:id', db.updateUser)
+  app.put('/users/:id', db.updateUserAsync)
+  //app.delete('/users/:id', db.deleteUser)
+  app.delete('/users/:id', db.deleteUserAsync)
+  
   //API of books
-  app.get('/books', db.getBooks)
+  // app.get('/books', db.getBooks)
+  app.get('/books', db.getBooksAsync)
   app.get('/books/:id', db.getBookByID)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 4]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
-  app.get('/categories', db.getCategory)
-  app.get('/categories/:id', db.getCategoryByID)
-  app.post('/categories', db.createCategory)
+  //app.get('/categories', db.getCategories)
+  app.get('/categories', db.getCategoriesAsync)
+  //app.get('/categories/:id', db.getCategoryByID)
+  app.get('/categories/:id', db.getCategoryByIDAsync)
+  //app.post('/categories', db.createCategory)
+  app.post('/categories', db.createCategoryAsync)
   app.put('/categories/:id', db.updateCategory)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 5]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
+  //app.put('/categories/:id', db.updateCategoryAsync)
   app.delete('/categories/:id', db.deleteCategory)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 6]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
+  //app.delete('/categories/:id', db.deleteCategoryAsync)
+
   //API of bookchecks
-  app.get('/bookchecks', db.getBookChecks) 
+  app.get('/bookchecks', db.getBookChecks)
+  //app.get('/bookchecks', db.getBookChecksAsync) 
   app.get('/bookchecks/:id', db.getBookChecksByID)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 7]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
-  app.post('/bookchecks', db.createBookCheck)
-  app.put('/bookchecks/:id', db.updateBookCheck)
+  //app.get('/bookchecks/:id', db.getBookcheckByIDAsync)
+  //app.post('/bookchecks', db.createBookCheck)
+  app.post('/bookchecks', db.createBookCheckAsync)
+  //app.put('/bookchecks/:id', db.updateBookCheck)
+  app.put('/bookchecks/:id', db.updateBookCheckAsync)
   app.delete('/bookchecks/:id', db.deleteBookCheck)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: index.js [hunk 8]
```diff
@@ -19,32 +19,62 @@ app.listen(port, () => {
...
+  //app.delete('/bookchecks/:id', db.deleteBookCheckAsync)
+
   //API of reports
-  app.get('/reports/numberofusers', db.getNumberOfUsersReport)
+  //app.get('/reports/numberofusers', db.getNumberOfUsers)
+  app.get('/reports/numberofusers', db.getNumberOfUsersAsync)
   app.get('/reports/numberofusersbygender', db.getNumberOfUsersByGender)
```
Comment (phkhuong) [inline index.js]: Please use async method for api

File: tests/test-users-api.js [hunk 9]
```diff
@@ -29,20 +29,27 @@ const cmnd = '82764619'
...
     before (async() => {
-        let result = await pool.query('SELECT public."InsertUser"($1, $2, $3, $4, $5, $6, $7) as new_id', 
-            [birthdate, signupdate, name, cmnd, email, gender, expirydate])
-        new_user_id = result.rows[0].new_id
-        console.log('id: ' + new_user_id)
+        // let result = await pool.query('SELECT public."InsertUser"($1, $2, $3, $4, $5, $6, $7) as new_id', 
+        //     [birthdate, signupdate, name, cmnd, email, gender, expirydate])
+        // new_user_id = result.rows[0].new_id
+        // console.log('id: ' + new_user_id)
+        try {
```
Comment (phkhuong) [inline tests/test-users-api.js]: Why should we need to use try/catch block here?

**Review by phkhuong — CHANGES_REQUESTED** [review]

File: tests/test-users-api.js [hunk 10]
```diff
@@ -93,7 +100,7 @@ describe("users api", async() => {
                 signupdate: signupdate.toLocaleDateString(),
                 expirydate: expirydate.toLocaleDateString()
             })
-            const result = pool.query('SELECT * FROM users WHERE id = $1', [response.data.InsertedUserId], (err, result) => {
+            const result = dbusers.get_user_by_id(response.data.InsertedUserId, (err, result) => {
                 if(err) {
```
Comment (phkhuong) [inline tests/test-users-api.js]: Please use async method

File: tests/test-users-api.js [hunk 11]
```diff
@@ -29,20 +29,27 @@ const cmnd = '82764619'
...
+            console.error(error)
+        }
+
     })
     describe('get users', async() => {
         it("1. should return status code 200 and the number of data response is equal the number of users in database when called correctly", async () => {
             let response = await axios.get( `${process.env.API_URL}/users`);
             let apiUsersCount = response.data.length;
-            pool.query('SELECT COUNT(1) FROM users', (error,res) => {
+            dbusers.count_number_of_users((error, result) => {
```
Comment (phkhuong) [inline tests/test-users-api.js]: Please use async method

File: index.js (hunk 1 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 2 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 3 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 4 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 5 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 6 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 7 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: index.js (hunk 8 above)
Comment (uyen3893) [reply index.js]: I fixed it

File: tests/test-users-api.js (hunk 9 above)
Comment (uyen3893) [reply tests/test-users-api.js]: I fixed it

File: tests/test-users-api.js (hunk 10 above)
Comment (uyen3893) [reply tests/test-users-api.js]: I fixed it

File: tests/test-users-api.js (hunk 11 above)
Comment (uyen3893) [reply tests/test-users-api.js]: I fixed it

**Review by phkhuong — APPROVED** [review]

**Pull request merged**